File: AUGUST-2025/13-08-2025/p326.erl
-spec is_power_of_three(N :: integer()) -> boolean().
is_power_of_three(N) -> inner(N, 0).

-spec inner(N :: integer(), R :: integer()) -> boolean().
inner(_, R) when R =/= 0 -> false;
inner(N, 0) when N < 1 -> false;
inner(1, 0) -> true;
inner(N, _) -> inner(N div 3, N rem 3).

File: AUGUST-2025/13-08-2025/p326.ex
defmodule Solution do
  @spec is_power_of_three(n :: integer) :: boolean
  def is_power_of_three(n), do: inner(n, 0)

  @spec inner(n :: integer, r:: integer) :: boolean
  defp inner(_, r) when r != 0, do: false
  defp inner(n, _) when n < 1, do: false
  defp inner(1, 0), do: true
  defp inner(n, _), do: inner(div(n, 3), rem(n, 3))
end